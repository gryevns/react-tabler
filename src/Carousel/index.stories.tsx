import React from "react";

import Carousel from "../Carousel";
import { boolean, number, withKnobs } from "@storybook/addon-knobs";

export default {
    title: "Carousel",
    decorators: [withKnobs],
};

export const Default = () => (
    <Carousel
        slide={true}
        indicators={boolean("Indicators", true)}
        controls={boolean("Controls", true)}
        interval={number("Interval", 2000)}
    >
        <Carousel.Item key="first" src="https://picsum.photos/id/221/1440/400">
            <h3>First slide label</h3>
            <p>Nulla vitae elit libero, a pharetra augue mollis interdum.</p>
        </Carousel.Item>
        <Carousel.Item key="second" src="https://picsum.photos/id/232/1440/400">
            <h3>Second slide label</h3>
            <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p>
        </Carousel.Item>
        <Carousel.Item key="third" src="https://picsum.photos/id/229/1440/400">
            <h3>Third slide label</h3>
            <p>Praesent commodo cursus magna, vel scelerisque nisl consectetur.</p>
        </Carousel.Item>
    </Carousel>
);
