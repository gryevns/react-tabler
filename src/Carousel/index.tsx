import React from "react";
import cn from "classnames";

export interface CarouselItemProps {
    active?: boolean;
    children?: React.ReactNode;
    src: string;
    style?: React.CSSProperties;
}

const CarouselItem = ({ active, children, src, style }: CarouselItemProps) => {
    const classes = cn("carousel-item", { active });
    return (
        <div className={classes}>
            <img
                className="d-block w-100"
                src={src}
                data-holder-rendered="true"
                style={style}
            />
            <div className="carousel-caption">{children}</div>
        </div>
    );
};

export interface CarouselProps {
    children: React.ReactNode[];
    interval?: number;
    slide?: boolean;
    controls?: boolean;
    indicators?: boolean;
}

const Carousel = ({
    children,
    interval = 5000,
    slide = true,
    controls = true,
    indicators = true,
}: CarouselProps) => {
    const [current, setCurrent] = React.useState(0);
    const slides = children.length;
    const previous = current === 0 ? slides - 1 : current - 1;
    const next = current === slides - 1 ? 0 : current + 1;
    const indicatorChildren = Array.from(Array(slides).keys()).map(i => {
        const props = {
            className: current === i ? "active" : "",
            onClick: () => setCurrent(i),
        };
        return <li key={`indicator-${i}`} {...props} />;
    });
    if (slide) {
        React.useEffect(() => {
            const slideDelay = setInterval(() => setCurrent(next), interval);
            return () => {
                clearTimeout(slideDelay);
            };
        });
    }
    const onNext = (e: React.MouseEvent) => {
        e.preventDefault();
        setCurrent(next);
    };
    const onPrevious = (e: React.MouseEvent) => {
        e.preventDefault();
        setCurrent(previous);
    };
    return (
        <div className="carousel slide" data-ride="carousel">
            {indicators && (
                <ol className="carousel-indicators">{indicatorChildren}</ol>
            )}
            <div className="carousel-inner">
                {children.map((child, i) =>
                    React.cloneElement(child as React.ReactElement<any>, {
                        active: i === current,
                    }),
                )}
            </div>
            {controls && (
                <>
                    <a
                        className="carousel-control-prev"
                        role="button"
                        href="#"
                        onClick={onPrevious}
                    >
                        <span
                            aria-hidden="true"
                            className="carousel-control-prev-icon"
                        ></span>
                        <span className="sr-only">Previous</span>
                    </a>
                    <a
                        className="carousel-control-next"
                        role="button"
                        href="#"
                        onClick={onNext}
                    >
                        <span
                            aria-hidden="true"
                            className="carousel-control-next-icon"
                        ></span>
                        <span className="sr-only">Next</span>
                    </a>
                </>
            )}
        </div>
    );
};

Carousel.Item = CarouselItem;

export default Carousel;
