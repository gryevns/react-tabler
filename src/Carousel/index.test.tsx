import React from "react";
import { shallow } from "enzyme";

import Carousel from "./index";

describe("Carousel", () => {
    it("should match types snapshot", () => {
        const component = shallow(
            <Carousel>
                <Carousel.Item key="first" src="first">
                    <h3>First slide label</h3>
                    <p>Nulla vitae elit libero, a pharetra augue mollis interdum.</p>
                </Carousel.Item>
                <Carousel.Item key="second" src="second">
                    <h3>Second slide label</h3>
                    <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit.</p>
                </Carousel.Item>
            </Carousel>,
        );
        expect(component).toMatchSnapshot();
    });
    it.skip("should iterate items after timeout", () => {
        const wrapper = shallow(
            <Carousel interval={1000}>
                <Carousel.Item key="first" src="first">
                    <h3>First slide label</h3>
                </Carousel.Item>
                <Carousel.Item key="second" src="second">
                    <h3>Second slide label</h3>
                </Carousel.Item>
            </Carousel>,
        );
        jest.useFakeTimers();
        setTimeout(() => {
            const active = wrapper.findWhere(
                n => n.name() === "CarouselItem" && n.prop("active") === true,
            );
            expect(active.prop("src")).toEqual("second");
        }, 1500);
        jest.runAllTimers();
    });
});
