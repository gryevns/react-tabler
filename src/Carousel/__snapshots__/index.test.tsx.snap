// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should match types snapshot 1`] = `
<div
  className="carousel slide"
  data-ride="carousel"
>
  <ol
    className="carousel-indicators"
  >
    <li
      className="active"
      key="indicator-0"
      onClick={[Function]}
    />
    <li
      className=""
      key="indicator-1"
      onClick={[Function]}
    />
  </ol>
  <div
    className="carousel-inner"
  >
    <CarouselItem
      active={true}
      key="first"
      src="first"
    >
      <h3>
        First slide label
      </h3>
      <p>
        Nulla vitae elit libero, a pharetra augue mollis interdum.
      </p>
    </CarouselItem>
    <CarouselItem
      active={false}
      key="second"
      src="second"
    >
      <h3>
        Second slide label
      </h3>
      <p>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit.
      </p>
    </CarouselItem>
  </div>
  <a
    className="carousel-control-prev"
    href="#"
    onClick={[Function]}
    role="button"
  >
    <span
      aria-hidden="true"
      className="carousel-control-prev-icon"
    />
    <span
      className="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    className="carousel-control-next"
    href="#"
    onClick={[Function]}
    role="button"
  >
    <span
      aria-hidden="true"
      className="carousel-control-next-icon"
    />
    <span
      className="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;
